### ISSUE TITLE
ProjectTo selects all columns when having nullable FK and FKName property

### ISSUE THREAD

**Issue by borisdj** [AutoMapper/AutoMapper#2183]
### Source/destination types
```c#
    public class User
    {
        public int UserId { get; set; }
    
        public string Email { get; set; }
    
        public string Name { get; set; }

        public int? PartyId { get; set; }
    
        public string Remark { get; set; } // this should NOT be in SELECT

        ...
    }

    public class UserModel
    {
        public int UserId { get; set; }

        public string Email { get; set; }
    
        public string Name { get; set; }

        public int? PartyId { get; set; }
        public string PartyName { get; set; } // JOIN by convention TableName = Table.Name
    }
```

### Mapping configuration

```c#
CreateMap<User, UserModel>();
```

### Version: 6.0.2
.NetCore app with EFCore.SqlServer 1.1.2

### Expected behavior
Select only columns from Model.
```sql
SELECT [u].[UserId], [u].[Email], [u].[Name], [u].[PartyId], [s].[Name]
FROM [dbo].[User] AS [u]
LEFT JOIN [dbo].[Party] AS [s] ON [u].[PartyId] = [s].[PartyId]
```
### Actual behavior
All columns from both tables are selected.
```sql
SELECT [u].[UserId], [u].[Email], [u].[Name], [u].[Remark], [u].[PartyId], [s].[PartyId], [s].[Name], [s].[Description]
FROM [dbo].[User] AS [u]
LEFT JOIN [dbo].[Party] AS [s] ON [u].[PartyId] = [s].[PartyId]
```
### Steps to reproduce

```csharp
var list = context.Set<User>().ProjectTo<UserModel>().ToList();
```

**Comment by lbargaoanu** [comment]
Does it work with EF6?

**Comment by borisdj** [comment]
I haven't check that, only for EFCore.

**Comment by lbargaoanu** [comment]
We don't fix EF core issues at this time. Report back if you can repro it with EF6.

**Comment by borisdj** [comment]
I just remembered had similar [issue](https://github.com/aspnet/EntityFramework/issues/8191) with EFCore itself.
So this might be the caused form the EFCore.
And that issue is said to be fixed in EFCore 2.0.0-preview1.
So I will check with this preview verison, maybe it's already fixed, if not we can discuss it more.

Will EFCore be included in active support after EFCore 2.0.0 release, which is planed for Q3 2017?

**Comment by lbargaoanu** [comment]
That's for @jbogard.

**Comment by jbogard** [comment]
I wonder if at some point we'll need different LINQ projecters for
different providers. If so this code might have to partially move to the
EF6 package and we'd make another for EFCore. But that'd be complicated.

**Comment by TylerCarlson1** [comment]
oh wouldn't another idea might work where you look at the IQueryProvider and based on the type determine to use EF6, EFCore, memory mapping, ect then make the query generate based off that?  You wouldn't need to reference EF6/EFCore directly in AM; just identify by TypeInfo name.  That way it can use both and not have to deal with the complexity of moving reusable code outside of AutoMapper and work across dlls.  Even look for versioning to compare EFCore 1.x vs 2.x and account for functionality that way if you want to go that far.

Personally I don't like this way either, because you have to rely on TypeInfo being exactly correct and newer versions having issues that the older versions didn't have and things like that.  It does keep existing functionality in tack though for AM projection with EF6 without having to get a new dll.

Maybe for breaking down projection you can have rules for which expression generators to use for which query type like ICanGenerateExpression and that way filter out by IQueryProvider if this type is applicable and use it.  Then make EF6 and EFCore versions that only work with their stated queryprovider.

**Comment by lbargaoanu** [comment]
It seems easier to move the EF6 code elsewhere and let people work on it if they need it and for us to move forward with EF Core. They plan a version after the 2.0 release, so I guess that should be stable enough. But I'd say that depends on what MS does with these libraries. How much are they going to invest in each, and what the community's response will be. If MS can get away with not evolving EF6, then I guess we can too :)

**Comment by borisdj** [comment]
I can confirm that this issue was caused by EFCore bug(link is previous comment) as I assumed.
So it is fixed is EFCore 2.0 that was released 2 days ago.

Regarding the library compatibility and maintaining, I find separating projects and 2 nugget for EF6 and EFCore to be better solution in the long run.
For the moment it's still the same but later EF6 will probably have only bug fixes, while EFCore will be evolving with new features as well.
Also code will be much cleaner when separated, otherwise there would be a lot of **if**-s.
Some hybrid option would be to make one Shared Project with all that is the same and then have 2 more project that would implement specifics of these 2 framework versions.

Also .NetStandard 2.0 is now Recommended target for libraries since it is compatible with project targeting both EFCore2.0+ or NetFramework4.6.1+
